### PR TITLE
558 access token for private repositories

### LINKED ISSUE #558: Problem with private repositores - download of imported files leads to 403
When I use imports within a private repository, Hets fails to download the imported ontologies, because they are not accessible in a private repository:
```
wget http://ontohub.org/colore4/magma/semigroup.clif
--2014-01-11 19:51:07--  http://ontohub.org/colore4/magma/semigroup.clif
Auflösen des Hostnamen »ontohub.org (ontohub.org)«... 134.102.204.133
Verbindungsaufbau zu ontohub.org (ontohub.org)|134.102.204.133|:80... verbunden.
HTTP-Anforderung gesendet, warte auf Antwort... 403 Forbidden
2014-01-11 19:51:07 FEHLER 403: Forbidden.
```
Maybe a special password should be transfered in the URL?

### PR DESCRIPTION
**Pull request by eugenk**
This shall fix #558. As soon as hets (https://github.com/spechub/Hets/issues/1422) can handle it completely, this should work. Now, it is working for non-owl files whose file type can be detected by hets' magic.

After merging, don't forget to put this line into the crontab for the expired tokens to be deleted daily:
```
0 0  * * *  ~/current/script/rails runner 'AccessTokenDeletionWorker.new.perform'
```

### CODE REVIEW

**Comment by eugenk** [comment]
We always pass an access token in the params. It has to be omitted when there is no access token

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -40,6 +40,14 @@ class ApplicationController < ActionController::Base
 
   protected
 
+  def current_ability
+    @current_ability ||= Ability.new(current_user, params[:'access-token'])
+  end
+
+  def params_to_pass_on_redirect
+    {:'access-token' => params[:'access-token']}
```
Comment (eugenk) [inline app/controllers/application_controller.rb]: No. This behavior is actually deactivated:
```yml
Style/SpaceInsideHashLiteralBraces:
  Description: "Use spaces inside hash literal braces - or don't."
  Enabled: true
  EnforcedStyle: no_space
```

File: app/models/ability.rb [hunk 2]
```diff
@@ -15,7 +15,10 @@ def initialize(user)
         if subject.is_private
           subject.permission?(:reader, user) ||
           subject.permission?(:editor, user) ||
-          subject.permission?(:owner, user)
+          subject.permission?(:owner, user) ||
+          subject.access_tokens.unexpired.any? do |token|
```
Comment (eugenk) [inline app/models/ability.rb]: In this file: No thanks. This will be deleted soon, anyway.

**Comment by eugenk** [comment]
in pause because it needs some rewriting after #1269 is merged (Hets::Options needs to be replaces)

**Comment by eugenk** [comment]
This is now fully adjusted to the changes of interfacing with Hets. It is open for review again.

**Comment by eugenk** [comment]
Oh and I rebased this branch on staging. In this process, the `Hets::Options` related commits were skipped (not needed any more, replaced by the new `HetsOptions`/`ProveOptions`).
Those skipped commits were the only ones with conflicts. The staging version has removed all the Hets-CLI related code and the `Hets::Options` related code from this branch changed those removed lines (and added - now useless - lines in between).

The first new commit is
Move class methods to the top.	f38207c

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -40,6 +40,16 @@ class ApplicationController < ActionController::Base
 
   protected
 
+  def current_ability
+    @current_ability ||= Ability.new(current_user, params[:'access-token'])
+  end
+
+  def params_to_pass_on_redirect
+    new_params = {}
+    %i(access-token).each { |key| new_params[key] = params[key] if params[key] }
```
Comment (0robustus1) [inline app/controllers/application_controller.rb]: How about using `select`?

Comment (eugenk) [inline app/controllers/application_controller.rb]: Right. `select` is better :D

File: app/models/repository/access.rb [hunk 4]
```diff
@@ -4,6 +4,24 @@ module Repository::Access
   OPTIONS = %w[public_r public_rw private_r private_rw]
   DEFAULT_OPTION = OPTIONS[0]
 
+  def self.as_read_only(access)
+    access.split('_').first + '_r'
```
Comment (0robustus1) [inline app/models/repository/access.rb]: how about using a limit to split (e.g. `...split('_', 2)`) so optimizations can kick in? I know it's probably just a short string, but why not ;)

File: app/models/repository/access.rb [hunk 5]
```diff
@@ -4,6 +4,24 @@ module Repository::Access
...
+  end
+
+  def destroy_expired_access_tokens
+    access_tokens.select(&:expired?).map(&:destroy)
```
Comment (0robustus1) [inline app/models/repository/access.rb]: Why `map`?

File: lib/hets/hets_options.rb [hunk 6]
```diff
@@ -46,9 +47,19 @@ def remove_nil_fields
...
+        if @options[:'url-catalog'].blank?
+          @options.delete(:'url-catalog')
+        else
+          @options[:'url-catalog'] = @options[:'url-catalog'].join(',')
+        end
+      end
+    end
+
+    def prepare_access_token
+      if @options[:'access-token'].is_a?(AccessToken)
```
Comment (0robustus1) [inline lib/hets/hets_options.rb]: Why the check for the specific class? I would've assumed that it is only important if the value is non-nil.

File: lib/router_constraints.rb [hunk 7]
```diff
@@ -130,9 +130,18 @@ def matches?(request, path = nil)
 
 class RefIRIRouterConstraint < IRIRouterConstraint
   def matches?(request)
+    regex = %r{
```
Comment (0robustus1) [inline lib/router_constraints.rb]: why not use a constant?

File: lib/router_constraints.rb [hunk 8]
```diff
@@ -130,9 +130,18 @@ def matches?(request, path = nil)
...
+      |
+      \?([^=;&]+)[&;].*$ # splits a lone MMT-arg from normal query-string
+      |
+      \?.*$ # drops a the whole standard query-string
+    }x
     # remove the ref/:version_number portion from path
     path = Journey::Router::Utils.unescape_uri(request.original_fullpath).
-      sub(%r{\A/ref/\d+/}, '')
+      sub(%r{\A/ref/\d+/}, '').
+      sub(regex, '\1')
```
Comment (0robustus1) [inline lib/router_constraints.rb]: Is the MMT-check still current? (we have the `MMTRouterConstraint` now)

File: lib/hets/hets_options.rb (hunk 6 above)
Comment (eugenk) [inline lib/hets/hets_options.rb]: The `prepare_*` methods convert the objects to `String` for passing them to the Hets API. This if condition checks if it already has been converted.

File: lib/router_constraints.rb (hunk 8 above)
Comment (eugenk) [inline lib/router_constraints.rb]: You're right, it shouldn't be necessary anymore. Do we want to support access tokens with MMT-style URIs?

File: app/models/repository/access.rb (hunk 5 above)
Comment (eugenk) [inline app/models/repository/access.rb]: An old habit from using Haskell: "Want to do something with all the elements? Use map." ;)
I will change this to `each`.

File: lib/hets/hets_options.rb (hunk 6 above)
Comment (0robustus1) [inline lib/hets/hets_options.rb]: Yeah, but `to_s` is basically a noop on Strings, so you could just skip the altogether, couldn't you?

File: lib/router_constraints.rb (hunk 8 above)
Comment (0robustus1) [inline lib/router_constraints.rb]: I'm not sure, to be honest. Access-Tokens are an internal feature (not exposed to the public), and we only supply `/ref/version_number`-based loc/ids to that. So is there any reason to support MMT?

Comment (eugenk) [inline lib/router_constraints.rb]: That's what I thought, too. So I just remove the MMT-handling code above and do nothing more about MMT.

Comment (0robustus1) [inline lib/router_constraints.rb]: Alright.

**Comment by 0robustus1** [comment]
:+1:

**Pull request merged**